[CSP] Export the default directives to allow for easier overloading

AubreyHewes opened the pull request:
If you only want to change one of the, out of the box (default), CSP directives. You are expected to copy the current [`DEFAULT_DIRECTIVES`](https://github.com/helmetjs/helmet/blob/53e31738d4d23ea3879e5f2b189622e8d8348f71/middlewares/content-security-policy/index.ts#L20) and change them as your needs require.

This PR just exports the `DEFAULT_DIRECTIVES` ensuring an easier dev experience...

i.e. instead of having to do;

````typescript
helmet({
  contentSecurityPolicy: {
    directives: {
      // copied from DEFAULT_DIRECTIVES https://github.com/helmetjs/helmet/blob/53e31738d4d23ea3879e5f2b189622e8d8348f71/middlewares/content-security-policy/index.ts#L20
      "default-src": ["'self'"],
      "base-uri": ["'self'"],
      "block-all-mixed-content": [],
      "font-src": ["'self'", "https:", "data:"],
      "frame-ancestors": ["'self'"],
      "img-src": ["'self'", "data:"],
      "object-src": ["'none'"],
      "script-src": ["'self'", "'unsafe-inline'"], // <- this is non default
      "script-src-attr": ["'none'"],
      "style-src": ["'self'", "https:", "'unsafe-inline'"],
      "upgrade-insecure-requests": [],
    },
  },
})
````

We can now do something like the following:

Override a directive

````typescript
import { DEFAULT_DIRECTIVES } from "helmet/dist/middlewares/content-security-policy";

helmet({
  contentSecurityPolicy: {
    directives: {
      ...DEFAULT_DIRECTIVES,
      "script-src": ["'self'", "'unsafe-inline'"], // <- this is non default
    },
  },
})
````

Extend a directive

````typescript
import { DEFAULT_DIRECTIVES } from "helmet/dist/middlewares/content-security-policy";

helmet({
  contentSecurityPolicy: {
    directives: {
      ...DEFAULT_DIRECTIVES,
      "script-src": [
        ...DEFAULT_DIRECTIVES["script-src"], 
        "'unsafe-inline'" // <- this is non default
      ],
    },
  },
})
````

Due to the simplicity of the change I did not see the need to add tests.

Let me know if I have missed something and/or tests are required!


A possible enhancement would be to export the CSP defaults directly;

````typescript
import { CSP_DEFAULT_DIRECTIVES } from "helmet";
````

EvanHahn commented:
Thanks for making this pull request.

I think this is an interesting idea, but I have two reservations:

1. It won't be exported properly. I want to support CommonJS (so that things like `require('helmet')` work) and TypeScript exports, which means I use `module.exports` (you can see that in the code). If we wanted to export this, we'd need it to dangle off of the CSP function. I'd want to make sure that works in all cases.
2. It boxes me in a bit as a maintainer if I ever want to change this.
   - I can no longer change the way this structure is represented internally. For example, I've considered making the default directives a hard-coded string for performance. I couldn't do that if I merged this.
   - This is debatable, but I think I could release a new version of Helmet with a _less_ restrictive CSP without a breaking change. That might have unexpected behavior.

What do you think?

AubreyHewes commented:
Hey, thanks for the feedback.

 1. Ah yeah, I did not notice this, as I use typescript.
     1. So I could dangle a method of the default export;

      ````typescript
      contentSecurityPolicy.getDefaultDirectives = () => DEFAULT_DIRECTIVES;
      ````

2. As it is now a method you _could_ internally change the defaults to a string though then this method _would_ have to then translate them to the object structure for downstream easier overloading...

      Usage would then be like;

      ```typescript
      // TypeScript
      import helmet from "helmet";
      // CommonJS
      const helmet = require("helmet");

      helmet({
        contentSecurityPolicy: {
          directives: {
            ...helmet.contentSecurityPolicy.getDefaultDirectives();,
          }
        }
      });
      ````

I think that this solution addresses your issues, what do you think?

EvanHahn commented:
Agreed, this addresses my concern.

I always want to be very cautious about adding any new "surface area" to the API because I'll have to support it for the lifetime of `helmet@4` and possibly beyond, depending on how many people come to rely on it. It's not your idea specifically—I think this is pretty reasonable. My caution is around _any_ addition.

I'll think more about this and get back to you. If you can find other folks who want this feature to leave their thoughts, that would help.

AubreyHewes commented:
@EvanHahn Thanks for the feedback. I understand the need for not extending the public interface. 

Our requirement for this possible change was due to our internal code reviews not liking "copying" the defaults.

I agree that the community should decide, Maybe tag this as a slow cooker or something? ;-)

EvanHahn commented:
Good idea. I've created a "feedback wanted" label and added it here.

I've set myself a reminder to come back to this sometime in November.

A-Tokyo commented:
This would be quite useful, specially when dealing with smth like graphiql https://github.com/apollographql/apollo-server/issues/4648

EvanHahn commented:
I just renamed `master` to `main` which caused this PR to be automatically closed. I didn't realize this would happen—apologies.

However, I'm thinking of changing this to be a function, like this:

```js
helmet.contentSecurityPolicy.getDefaultDirectives()
```

That way, someone couldn't modify the object in some way and cause problems. Does that sound reasonable?

omeid commented:
The alternative is that instead of exporting the defaults, you could export a function that accepts the same parameters as the middleware but merges them with the default, so that you can do something like this:

```ts
const csp = require("helmet-csp");

app.use(
  csp.withDefaults({
    directives: {
      upgradeInsecureRequests: false,
    }
  })
);
```

`ContentSecurityPolicyOptions` is already part of the public API so changing that would be breaking changes anyways.

A-Tokyo commented:
@EvanHahn Sounds resonable!

EvanHahn commented:
I just set this up in #276. Please let me know what you think. If it looks good, I'll release it in an upcoming version of Helmet.

AubreyHewes commented:
@EvanHahn yeah seems good exactly what I proposed ;-)
